src/transformers/modeling_gguf_pytorch_utils.py

```diff
@@ -0,0 +1,163 @@
+
+    architecture = read_field(reader, "general.architecture")[0]
+    model_name = read_field(reader, "general.name")
+
+    if "llama" in architecture and "mistral" in model_name:
+        updated_architecture = "mistral"
+    else:
+        updated_architecture = architecture
+
+    if architecture not in GGUF_SUPPORTED_ARCHITECTURES:
+        raise ValueError(f"Architecture {architecture} not supported")
+
+    # List all key-value pairs in a columnized format
+    for gguf_key, field in reader.fields.items():
+        gguf_key = gguf_key.replace(architecture, updated_architecture)
+        split = gguf_key.split(".")
+        prefix = split[0]
+        config_key = ".".join(split[1:])
+
+        value = [_gguf_parse_value(field.parts[_data_index], field.types) for _data_index in field.data]
+
+        if len(value) == 1:
+            value = value[0]
+
+        if isinstance(value, str) and architecture in value:
+            value = value.replace(architecture, updated_architecture)
+
+        for parameter in GGUF_TO_TRANSFORMERS_MAPPING:
+            parameter_renames = GGUF_TO_TRANSFORMERS_MAPPING[parameter]
+            if prefix in parameter_renames and config_key in parameter_renames[prefix]:
+                renamed_config_key = parameter_renames.get(prefix, {}).get(config_key)
+                if renamed_config_key == -1:
+                    continue
+
+                if renamed_config_key is not None:
+                    parsed_parameters[parameter][renamed_config_key] = value
+
+                if gguf_key in reader_keys:
+                    reader_keys.remove(gguf_key)
+
+        if gguf_key in reader_keys:
+            logger.info(f"Some keys were not parsed and added into account {gguf_key} | {value}")
+
+    if return_tensors:
+        tensor_key_mapping = GGUF_TO_TRANSFORMERS_MAPPING["tensors"][architecture]
+
+        for tensor in tqdm(reader.tensors, desc="Converting and de-quantizing GGUF tensors..."):
+            renamed_tensor_name = tensor.name
+
+            for tensor_name_mapping in GGUF_TO_TRANSFORMERS_MAPPING["tensors"]:
+                if tensor_name_mapping in renamed_tensor_name:
+                    renamed_tensor_name = renamed_tensor_name.replace(
+                        tensor_name_mapping, GGUF_TO_TRANSFORMERS_MAPPING["tensors"][tensor_name_mapping]
+                    )
+
+            shape = tensor.shape
+            name = tensor.name
+
+            weights = load_dequant_gguf_tensor(shape=shape, ggml_type=tensor.tensor_type, data=tensor.data)
+
+            if architecture == "llama" and (".attn_k." in name or ".attn_q." in name):
+                num_heads = parsed_parameters["config"]["num_attention_heads"]
+                tmp_shape = (int(shape[-1] // num_heads // 2), num_heads, 2, shape[0])
+                weights = weights.reshape(tmp_shape)
+                weights = weights.transpose(0, 2, 1, 3)
+                weights = weights.reshape(shape[::-1])
```

Comment:
We'll want to make this more general when we have more models - a problem for future us!